2807: Labels now have a 0.5 outline

bnelson05 opened the pull request:
Resolves #2807 

Using ctx, I drew a black and white outline around the label icon image.

![2807Before](https://user-images.githubusercontent.com/69987726/178040050-c7e035c3-6749-43f9-b660-bb4e8b072cf5.JPG)
![2807After](https://user-images.githubusercontent.com/69987726/178040042-8eb19385-77c6-449e-b7f7-d7f1b2da3a9e.JPG)

##### Testing instructions
1. Place a label in the Explore tab, and it should have an outline.

misaugstad commented:
@bnelson05 this issue is actually about adding some sort of outline to all of the main places where users see labels: /audit, /validate, and /labelMap! I think that the one you've done already was most likely the hardest one though! :grin:

bnelson05 commented:
Here is the thinner border style (about 0.5px - 1px).
![2957GalleryThinOutline](https://user-images.githubusercontent.com/69987726/178600761-26db4faf-335a-42cc-b6f2-a01309b8fa8c.JPG)
![2957ValidateThinOutline](https://user-images.githubusercontent.com/69987726/178600763-8d70015a-42de-4234-a574-5773a7e569ef.JPG)

Here is the thicker border style (about 1.5px).
![2957GalleryThickOutline](https://user-images.githubusercontent.com/69987726/178600784-dd3e2338-55eb-4c5b-8154-2fcb8aeff5b0.JPG)
![2957ValidateThickOutline](https://user-images.githubusercontent.com/69987726/178600788-668b8b34-0534-4bee-b8e0-f5d2c8829ec3.JPG)

misaugstad commented:
I think I like the thinner outlines better almost always. I forget, have we tried swapping black and white as well?

misaugstad commented:
Could you actually show an example for each of the four main label types (curb ramp, missing ramp, obstacle, surface problem) with black on the outside and white on the outside?

misaugstad commented:
This looks really good so far!! Just a couple small visual updates to make.

First, in Gallery on the expanded view, the border is making the label a bit off center. But I realized that if you increase the width/height of the marker from 20 px to 22 px it looks okay (because the underlying image is 20 px wide). Here are pics before and after I made that change manually.
<img width="304" alt="Screen Shot 2022-07-13 at 12 22 57 PM" src="https://user-images.githubusercontent.com/6518824/178817388-79854e16-0055-4bdd-a653-e3cfcbbde18f.png">
<img width="304" alt="Screen Shot 2022-07-13 at 12 33 34 PM" src="https://user-images.githubusercontent.com/6518824/178817405-1ec819ec-1cc1-4464-b7ac-7a0e35295867.png">

The other issue is that on Validate we have a "hide label" feature that mostly hides a label so that users can see underneath it. The border is being kept when we hide the label, and it should probably be removed in that case. Here's how it looks on develop vs this branch rn.
<img width="304" alt="Screen Shot 2022-07-13 at 12 14 01 PM" src="https://user-images.githubusercontent.com/6518824/178817638-8c6462e7-aeb6-4c01-a659-835937de1486.png">
<img width="304" alt="Screen Shot 2022-07-13 at 12 14 07 PM" src="https://user-images.githubusercontent.com/6518824/178817649-1875c5e8-4582-4e1d-aac9-7cc496694859.png">

bnelson05 commented:
I have updated the width/height of the marker from 20px to 22px, and I added a CSS class called `icon-outline` that is used for hiding/unhiding the label as well as the Gallery.

misaugstad commented on public/javascripts/SVLabel/src/SVLabel/label/Label.js:
```diff
@@ -423,6 +423,16 @@ function Label (svl, pathIn, params) {
             // Renders the label image.
             path.render2(ctx, pov);
 
+            ctx.lineWidth = .7;
```
Can you add a comment above this block of code so that we know that this code draws the outline?

misaugstad commented on public/javascripts/SVValidate/src/label/LabelVisibilityControl.js:
```diff
@@ -32,6 +32,8 @@ function LabelVisibilityControl () {
         let panomarker = svv.panorama.getPanomarker();
         let label = svv.panorama.getCurrentLabel();
+        var marker = document.getElementById('validate-pano-marker');
+        marker.classList.add('icon-outline');
```
I think that you can probably replace these two lines with just
```
panomarker.marker_.classList.add('icon-outline');
```

We're already getting the panomarker here, so it's just about getting the actual HTML element out of it.

misaugstad replied:
If you do end up needing to save a variable, try to use `let` instead of `var` if possible!

misaugstad replied:
>I tried to combine the two lines in `LabelVisibilityControl.js` as you suggested but it didn't work the way I thought it would, so I just changed `var` to `let`

Interesting... I just tested it out locally while using your branch and it worked exactly the same! In fact, if I log both of these to the console, I get identical output:
```
console.log(document.getElementById('validate-pano-marker'));
console.log(panomarker.marker_);
```

misaugstad reviewed (changes requested):
A few more small visual things :) Pardon the nit-picking, we want this to look perfect, and it's so close!

I'm getting that same issue on /validate that I was getting on /gallery before, where the border is making the icon a bit off center. This was probably always happening on /validate, I just likely didn't happen to test with the obstacle label type where this is most obvious.
![Screenshot from 2022-08-01 16-28-43](https://user-images.githubusercontent.com/6518824/182262045-8d10d22d-f7af-440e-a2be-18423c0c2b0a.png)

And on /validate, this a bug was introduced in your code. If you hide a label and then validate it while it is hidden, the button is continues to say "show label" even though the new label is shown by default. Then if you click the button, the text switches back to "hide label" and the icon doesn't change. Then if you click the 'hide' button again, it hides.

The gist is that you now have to click the hide/show button twice if you validate a hidden label. Here it is in pictures!

Hide first label
![Screenshot from 2022-08-01 16-27-10](https://user-images.githubusercontent.com/6518824/182262303-4d768e90-a8d4-453a-971f-b033a6251fca.png)

Validate it and a new label appears (unhidden). But the button says "show label" instead of "hide label"
![Screenshot from 2022-08-01 16-27-32](https://user-images.githubusercontent.com/6518824/182262327-8459342e-f6e6-423e-b0a1-af264f407b0d.png)

After clicking "show label" it switches to saying "hide label"
![Screenshot from 2022-08-01 16-27-44](https://user-images.githubusercontent.com/6518824/182262380-8180c74b-33aa-402e-bb68-8923724568cd.png)

Now when you click "hide label" it is hidden
![Screenshot from 2022-08-01 16-27-52](https://user-images.githubusercontent.com/6518824/182262411-c47856bb-d8bd-4fae-af31-280491e3fb61.png)

bnelson05 commented:
I centered the obstacle label in Validate and fixed the issue with the hide/show button. I tried to combine the two lines in `LabelVisibilityControl.js` as you suggested but it didn't work the way I thought it would, so I just changed `var` to `let`.

misaugstad commented on public/javascripts/SVLabel/src/SVLabel/label/Label.js:
```diff
@@ -423,6 +423,17 @@ function Label (svl, pathIn, params) {
             // Renders the label image.
             path.render2(ctx, pov);
 
+            //Draws label outline
```
Can you modify the comment slightly to fit the [style guide](https://github.com/ProjectSidewalk/SidewalkWebpage/wiki/Style-Guide)?

misaugstad reviewed (changes requested):
We are so close, but there is one new bug!

Whenever you hide a label and then validate it while it's hidden, you will see this error in the console.
<img width="410" alt="Screen Shot 2022-08-08 at 9 38 06 AM" src="https://user-images.githubusercontent.com/6518824/183469905-a28fc61b-22b3-4585-be36-c9c6171b5bef.png">

This normally doesn't cause an issue, but if you do it on the last label in a mission, you are never able to start the next mission! The 'Validate more labels' button is remains grayed out and inactive.
<img width="1039" alt="Screen Shot 2022-08-08 at 9 36 37 AM" src="https://user-images.githubusercontent.com/6518824/183470048-2a8d605b-8b83-49cd-acb5-1e4c704b9e83.png">

bnelson05 commented:
The Uncaught TypeError bug should now be fixed.

misaugstad reviewed (approved):
Thanks for this amazing PR @bnelson05!!

In the interest of time given that it's your last week, I made a tiny update to the comments and removed some whitespace you accidentally added. You can see that commit [here](https://github.com/ProjectSidewalk/SidewalkWebpage/pull/2957/commits/7a40647025ea4b9a8c64ff77a7e8eaa5c1930bb3).

And then I added the outline to labels that I just added to the user dashboard last week, since that's something that you didn't have access to while you were writing this code! You can find that [here](https://github.com/ProjectSidewalk/SidewalkWebpage/pull/2957/commits/fbb4cdcbfa84a667c78e0c8f8f9d934188d040b7).
